clients/clients-integration-tests/src/test/java/org/apache/kafka/clients/TransactionsExpirationTest.java

```diff
@@ -0,0 +1,310 @@
+                assertTrue(oldProducerEpoch + 2 <= newProducerEpoch);
+            } else {
+                assertEquals(oldProducerEpoch + 1, newProducerEpoch);
+            }
+
+            assertConsumeRecords(clusterInstance, List.of(TOPIC1), 2);
+        }
+    }
+
+    private void waitUntilTransactionalStateExists(ClusterInstance clusterInstance) throws InterruptedException {
+        try (Admin admin = clusterInstance.admin()) {
+            TestUtils.waitForCondition(() -> {
+                try {
+                    admin.describeTransactions(List.of(TRANSACTION_ID)).description(TRANSACTION_ID).get();
+                    return true;
+                } catch (Exception e) {
+                    return false;
+                }
+            }, "Transactional state was never added.");
+        }
+    }
+
+    private void waitUntilTransactionalStateExpires(ClusterInstance clusterInstance) throws InterruptedException {
+        try (Admin admin = clusterInstance.admin()) {
+            TestUtils.waitForCondition(() -> {
+                try {
+                    admin.describeTransactions(List.of(TRANSACTION_ID)).description(TRANSACTION_ID).get();
+                    return false;
+                } catch (Exception e) {
+                    return e.getCause() instanceof TransactionalIdNotFoundException;
+                }
+            }, "Transaction state never expired.");
+        }
+    }
+
+    private List<ProducerState> producerState(ClusterInstance clusterInstance) throws ExecutionException, InterruptedException {
+        try (Admin admin = clusterInstance.admin()) {
+            return admin.describeProducers(List.of(TOPIC1_PARTITION0)).partitionResult(TOPIC1_PARTITION0).get().activeProducers();
+        }
+    }
+
+    private void assertConsumeRecords(
+        ClusterInstance clusterInstance,
+        List<String> topics,
+        int expectedCount
+    ) throws InterruptedException {
+        for (GroupProtocol groupProtocol : clusterInstance.supportedGroupProtocols()) {
+            ArrayList<ConsumerRecord<byte[], byte[]>> consumerRecords = new ArrayList<>();
+            try (Consumer<byte[], byte[]> consumer = clusterInstance.consumer(Map.of(
+                    ConsumerConfig.GROUP_PROTOCOL_CONFIG, groupProtocol.name(),
+                    ConsumerConfig.ENABLE_AUTO_COMMIT_CONFIG, "false",
+                    ConsumerConfig.ISOLATION_LEVEL_CONFIG, "read_committed"
+                )
+            )) {
+                consumer.subscribe(topics);
+                TestUtils.waitForCondition(() -> {
+                    ConsumerRecords<byte[], byte[]> records = consumer.poll(Duration.ofMillis(100));
+                    records.forEach(consumerRecords::add);
+                    return consumerRecords.size() == expectedCount;
+                }, "Consumer with protocol " + groupProtocol.name + " should consume " + expectedCount + " records, but get " + consumerRecords.size());
```

Comment:
` consumerRecords.size()` is always zero, since the string is not lazily evaluated.

Reply:
Updated it. Thanks.